File: src/ros2_bridge/mappers/control_msgs/multi_dof_state_stamped.rs
//! Typed mapper for `control_msgs/msg/MultiDOFStateStamped`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn multi_dof_state_stamped_to_bus(msg: ros_env::control_msgs::msg::MultiDOFStateStamped) -> crate::control_msgs::msg::v1::MultiDofStateStamped {
    crate::control_msgs::msg::v1::MultiDofStateStamped {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        dof_states: msg.dof_states.into_iter().map(crate::ros2_bridge::mappers::control_msgs::single_dof_state::single_dof_state_to_bus).collect(),
    }
}

pub(crate) fn multi_dof_state_stamped_to_ros(bus: crate::control_msgs::msg::v1::MultiDofStateStamped) -> ros_env::control_msgs::msg::MultiDOFStateStamped {
    ros_env::control_msgs::msg::MultiDOFStateStamped {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        dof_states: bus.dof_states.into_iter().map(crate::ros2_bridge::mappers::control_msgs::single_dof_state::single_dof_state_to_ros).collect(),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ControlMsgsMultiDofStateStampedMapper;

impl TypedTopicMapper for ControlMsgsMultiDofStateStampedMapper {
    type Ros = ros_env::control_msgs::msg::MultiDOFStateStamped;
    type Bus = crate::control_msgs::msg::v1::MultiDofStateStamped;

    fn type_name(&self) -> &'static str {
        "control_msgs/msg/MultiDOFStateStamped"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(multi_dof_state_stamped_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(multi_dof_state_stamped_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/control_msgs/multi_dof_state_stamped.py
"""Generated mapper for `control_msgs/msg/MultiDOFStateStamped`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.control_msgs.single_dof_state import single_dof_state_to_bus, single_dof_state_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.control_msgs.msg.v1 import MultiDOFStateStamped as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def multi_dof_state_stamped_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.dof_states.extend([single_dof_state_to_bus(x) for x in msg.dof_states])
    return bus


def multi_dof_state_stamped_to_ros(bus):
    from control_msgs.msg import MultiDOFStateStamped as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.dof_states = [single_dof_state_to_ros(x) for x in bus.dof_states]
    return out


class ControlMsgsMultiDofStateStampedMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from control_msgs.msg import MultiDOFStateStamped as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return multi_dof_state_stamped_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return multi_dof_state_stamped_to_ros(bus)
